consensus/istanbul/backend/backend.go

```diff
@@ -497,9 +498,27 @@ func (sb *Backend) Commit(proposal istanbul.Proposal, aggregatedSeal types.Istan
 		return nil
 	}
 
-	if sb.broadcaster != nil {
-		sb.broadcaster.Enqueue(fetcherID, block)
+	// If no cached result, we import block through fetcher, which requires processing the block one more time.
+	// Else if cache hit, we directly write the block.
+	if result == nil {
+		sb.logger.Warn("state process cache miss, enqueuing the block to fetcher", "number", block.Number(), "hash", block.Hash())
+		if sb.broadcaster != nil {
+			sb.broadcaster.Enqueue(fetcherID, block)
+		}
+		sb.logger.Warn("Cannot import consensus block", "number", block.Number(), "hash", block.Hash())
+		return nil
+	}
+
+	if block := sb.chain.GetBlock(block.Hash(), block.NumberU64()); block != nil {
+		sb.logger.Info("Duplicate block to commit", "number", block.Number(), "hash", block.Hash())
+		return nil
+	}
+	if err := sb.writeBlockWithState(block, result.Receipts, result.Logs, result.State, true); err != nil {
+		sb.logger.Error("Failed writeBlockWithState", "err", err)
+		return err
```

Comment:
This seems like it is going to be racy, between the call to `GetBlock` and the call to `writeBlockWithState` a block may have been written.

So I think we could just skip the call to `GetBlock`. It would be good to be able to tell from the error returned from `writeBlockWithState` if the block was a duplicate or if it was an inconsistent block (same height but different hash). I think inconsistent blocks should never happen (I.E it implies consensus has been broken) so I guess we would want to stop processing in this case.

It looks like at present inconsistent blocks are ignored in write block with state.
https://github.com/celo-org/celo-blockchain/blob/ebd9817f69659a1ce9731f5aef60a023d9d09a7f/core/blockchain.go#L1304-L1306

So I think if an inconsistent block were encountered it would just overwrite the pre-existing block at that height, which doesn't seem like a good approach.